src/DistributedComputations/distributed_transpose.jl

```diff
@@ -0,0 +1,193 @@
+    i, j, k = @index(Global, NTuple)
+    size = N[1], N[2], n[3]
+    @inbounds begin
+        k′  = mod(k - 1, size[3]) + 1
+        m   = (k - 1) ÷ size[3]
+        idx = k′ + size[3] * (i-1 + size[1] * (j-1)) + m * prod(size)
+        zfield[i, j, k] = yzbuff.recv[idx]
+    end
+end
+
+# unpacking a y buffer from a communication with z-local direction
+@kernel function _unpack_buffer_yz!(yzbuff, yfield, N, n) # z -> y
+    i, j, k = @index(Global, NTuple)
+    size = N[1], n[2], N[3]
+    @inbounds begin
+        j′  = mod(j - 1, size[2]) + 1
+        m   = (j - 1) ÷ size[2]
+        idx = j′ + size[2] * (i-1 + size[1] * (k-1)) + m * prod(size)
+        yfield[i, j, k] = yzbuff.recv[idx]
+    end
+end
+
+# unpacking a y buffer from a communication with x-local direction
+@kernel function _unpack_buffer_yx!(yzbuff, yfield, N, n) # x -> y
+    i, j, k = @index(Global, NTuple)
+    size = N[1], n[2], N[3]
+    @inbounds begin
+        j′  = mod(j - 1, size[2]) + 1
+        m   = (j - 1) ÷ size[2] 
+        idx = j′ + size[2] * (k-1 + size[3] * (i-1)) + m * prod(size)
+        yfield[i, j, k] = yzbuff.recv[idx]
+    end
+end
+
+pack_buffer_x!(buff, f)  = launch!(architecture(f), f.grid, :xyz, _pack_buffer_x!,  buff, f, size(f))
+pack_buffer_z!(buff, f)  = launch!(architecture(f), f.grid, :xyz, _pack_buffer_z!,  buff, f, size(f))
+pack_buffer_yx!(buff, f) = launch!(architecture(f), f.grid, :xyz, _pack_buffer_yx!, buff, f, size(f))
+pack_buffer_yz!(buff, f) = launch!(architecture(f), f.grid, :xyz, _pack_buffer_yz!, buff, f, size(f))
+
+unpack_buffer_x!(f, fo, buff)  = launch!(architecture(f), f.grid, :xyz, _unpack_buffer_x!,  buff, f, size(f), size(fo))
+unpack_buffer_z!(f, fo, buff)  = launch!(architecture(f), f.grid, :xyz, _unpack_buffer_z!,  buff, f, size(f), size(fo))
+unpack_buffer_yx!(f, fo, buff) = launch!(architecture(f), f.grid, :xyz, _unpack_buffer_yx!, buff, f, size(f), size(fo))
+unpack_buffer_yz!(f, fo, buff) = launch!(architecture(f), f.grid, :xyz, _unpack_buffer_yz!, buff, f, size(f), size(fo))
+
+for (from, to, buff) in zip([:y, :z, :y, :x], [:z, :y, :x, :y], [:yz, :yz, :xy, :xy])
+    transpose!      = Symbol(:transpose_, from, :_to_, to, :(!))
+    pack_buffer!    = from == :y ? Symbol(:pack_buffer_, from, to, :(!)) : Symbol(:pack_buffer_, from, :(!))
+    unpack_buffer!  = to == :y ? Symbol(:unpack_buffer_, to, from, :(!)) : Symbol(:unpack_buffer_, to, :(!))
+    
+    buffer = Symbol(buff, :buff)
+    fromfield = Symbol(from, :field)
+    tofield = Symbol(to, :field)
+
+    transpose_name = string(transpose!)
+    to_name = string(to)
+    from_name = string(from)
+
+    pack_buffer_name = string(pack_buffer!)
+    unpack_buffer_name = string(unpack_buffer!)
+
+    @eval begin
+        """
+            $($transpose_name)(pf::TransposableField)
+
+        Transpose the fields in `TransposableField` from a $($from_name)-local configuration
+        (located in `pf.$($from_name)field`) to a $($to_name)-local configuration located
```

Comment:
When you say "local", does it mean that it's a continuous direction? i.e. not partitioned into different GPUs?

Reply:
yeah the the direction is "local" in each rank (every rank contains the whole direction)